minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,796 @@
+        return (self._ccoord.k is None) and (self._shape.t is None)
+
+    def to_quo(self) -> ZNode:
+        """Returns the quotient ZNode corresponding to self"""
+        qshape = ZShape(m=self._shape.m)
+        qccoord = CartesianCoord(x=self._ccoord, y=self._ccoord)
+        return ZNode(coord=qccoord, shape=qshape, convert_to_z=self.convert_to_z)
+
+    def is_vertical(self) -> bool:
+        """Returns True if the node represents a vertical qubit (i.e., its `u` coordinate in Zephyr coordinates is 0)."""
+        return self.node_kind is NodeKind.VERTICAL
+
+    def is_horizontal(self) -> bool:
+        """Returns True if the node represents a horizontal qubit (i.e., its `u` coordinate in Zephyr coordinates is 1)."""
+        return self.node_kind is NodeKind.HORIZONTAL
+
+    def neighbor_kind(
+        self,
+        other: ZNode,
+    ) -> EdgeKind | None:
+        """Returns the kind of edge between self and other;
+        EdgeKind if there is an edge in perfect Zephyr between them or None.
+
+        Args:
+            other (ZNode): a ZNode
+
+        Returns:
+            EdgeKind | None: The edge kind between self and other in perfect Zephyr or None if there is no edge between in perfect Zephyr.
+        """
+        if not isinstance(other, ZNode):
+            other = ZNode(other)
+        if self._shape != other._shape:
+            return None
+        coord1 = self._ccoord
+        coord2 = other._ccoord
+        x1, y1 = coord1.x, coord1.y
+        x2, y2 = coord2.x, coord2.y
+        if abs(x1 - x2) == abs(y1 - y2) == 1:
+            return EdgeKind.INTERNAL
+        if x1 % 2 != x2 % 2:
+            return None
+        if coord1.k != coord2.k:  # odd, external neighbors only on the same k
+            return None
+        if self.node_kind is NodeKind.VERTICAL:  # self vertical
+            if x1 != x2:  # odd, external neighbors only on the same vertical lines
+                return None
+            diff_y = abs(y1 - y2)
+            return EdgeKind.ODD if diff_y == 2 else EdgeKind.EXTERNAL if diff_y == 4 else None
+
+        # else, self is horizontal
+        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+            return None
+        diff_x = abs(x1 - x2)
+        return EdgeKind.ODD if diff_x == 2 else EdgeKind.EXTERNAL if diff_x == 4 else None
+
+    def internal_neighbors_generator(
+        self,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> Generator[ZNode]:
+        """Generator of internal neighbors of self when restricted by `where`.
+
+        Args:
+            where (Callable[[CartesianCoord | ZephyrCoord], bool], optional):
+                A coordinate filter. Applies to `ccoord` if `self.convert_to_z` is False,
+                or to `zcoord` if `self.convert_to_z` is True. Defaults to `lambda coord: True`.
+
+        Yields:
+            ZNode: Internal neighbors of self when restricted by `where`.
```

Comment:
Don't forget that sphinx docstrings should have double back-ticks (``` ``...`` ```) for code words and single back-ticks with roles for classes/methods/attributes, etc. (``:class:`...` ``)
```suggestion
        """Generator of internal neighbors of self when restricted by ``where``.

        Args:
            where (Callable[[CartesianCoord | ZephyrCoord, bool]]):
                A coordinate filter. Applies to ``ccoord`` if ``self.convert_to_z`` is ``False``,
                or to ``zcoord`` if :py:attr:`self.convert_to_z` is ``True``. Defaults to always 

        Yields:
            ZNode: Internal neighbors of self when restricted by `where`.
```

Reply:
Partially done in cc2e009a0484d628a96e0ef32748dbbb4c286fbb

Reply:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561